docs/reactive-nonblocking.qmd

```diff
@@ -0,0 +1,316 @@
+---
+title: Non-blocking operations
+---
+
+Sometimes in a Shiny app, you need to perform a long-running operation, like loading a large dataset or doing an expensive computation.
+If you do this in a reactive context, it will block the rest of the application from running until the operation completes.
+This can be frustrating for users, who may think that the app has crashed.
+
+Worse, if you have multiple users, then one user's long-running operation will block the other users' apps from running as well.
+These other users will not even be aware that their slow performance is due to another user's actions.
+
+In this article, we'll look at how to make your Shiny apps more responsive by using non-blocking operations.
+We'll also go out of our way to explain why the usual Python async techniques don't work the way you expect in Shiny.
+
+## Using async/await in Shiny
+
+Asynchronous programming is a technique used in Python to increase scalability and responsiveness, usually in networked apps.
+Python supports async programming at the language level, using the `async`/`await` keywords; in the standard library, with the [`asyncio`](https://docs.python.org/3/library/asyncio.html) module; and in third-party libraries, like [FastAPI](https://fastapi.tiangolo.com/) and [aiohttp](https://docs.aiohttp.org/).
+
+Shiny has async support as well, but it's a bit different from your typical async Python framework.
+On the one hand, Shiny is built on top of [Starlette](https://www.starlette.io/), which is an async web framework, so it's possible to use async functions in many parts of your Shiny app. On the other hand, Shiny is also designed around reactive programming concepts, and that creates constraints on how async functions can be used.
+
+### Reactive async/await
+
+Shiny supports the use of `async` and `await` in your reactive code.
+You can use `async` functions in `@reactive.effect`, `@reactive.calc`, and render functions, and within those async functions you can use `await` to wait for the results of other async functions.
+
+```{shinylive-python}
+#| standalone: true
+#| components: [editor, viewer]
+import asyncio
+import datetime
+from random import random
+import time
+from shiny import App, reactive, render, ui
+
+app_ui = ui.page_fluid(
+    ui.p("The time is ", ui.output_text("current_time", inline=True)),
+    ui.hr(),
+    ui.input_numeric("x", "x", value=1),
+    ui.input_numeric("y", "y", value=2),
+    ui.input_task_button("btn", "Add numbers"),
+    ui.output_text("sum"),
+)
+
+def server(input, output, session):
+    @render.text
+    def current_time():
+        reactive.invalidate_later(1)
+        return datetime.datetime.now().strftime("%H:%M:%S %p")
+
+    @reactive.calc
+    @reactive.event(input.btn)
+    async def sum_values():
+        await asyncio.sleep(5)
+        return input.x() + input.y()
+
+    @render.text
+    async def sum():
+        return str(await sum_values())
+
+app = App(app_ui, server)
```

Comment:
What's your reasoning for just including Core apps? Would you prefer to have a Express / Core tab option?

Reply:
This was intended to go live before Express, it's supposed to merge into main not rc-express. Although the window keeps shrinking...

Reply:
We'll need to come back to this before merging rc-express (if we don't, it'll feel awkward that everything else in reactivity and before is Express first, but this is Core first) 

So, ideally, I'd prefer to do those changes now and merge into rc-express. AFAICT, there is nothing in the critical path of the content there (other than the API reference, which doesn't seem absolutely critical, and should get done this week).